src/transformers/models/llama/modeling_llama.py

```diff
@@ -378,16 +380,12 @@ def forward(
 
         kv_seq_len = key_states.shape[-2]
         if past_key_value is not None:
-            kv_seq_len += past_key_value[0].shape[-2]
+            kv_seq_len += past_key_value.get_seq_length(self.layer_idx)
         cos, sin = self.rotary_emb(value_states, seq_len=kv_seq_len)
         query_states, key_states = apply_rotary_pos_emb(query_states, key_states, cos, sin, position_ids)
 
         if past_key_value is not None:
-            # reuse k, v, self_attention
-            key_states = torch.cat([past_key_value[0], key_states], dim=2)
-            value_states = torch.cat([past_key_value[1], value_states], dim=2)
-
-        past_key_value = (key_states, value_states) if use_cache else None
+            key_states, value_states = past_key_value.update(key_states, value_states, self.layer_idx, cos, sin)
```

Comment:
Nice! I like the idea of passing sin and cos here -> Are we sure that this doesn't lead to slow downs though?

Reply:
It should not, see: https://github.com/huggingface/transformers/pull/26681/files/e4c072d53f9b275581c9e20dd1ce474e74bbf391#r1392564863

But I will try to incorporate some naive latency tests in my performance experiments today/tomorrow.